core/lib/zksync_core_leftovers/src/lib.rs

```diff
@@ -741,6 +748,38 @@ pub async fn initialize_components(
         .context("add_tee_verifier_input_producer_to_task_futures()")?;
     }
 
+    if eth
+        .sender
+        .clone()
+        .context("eth_sender")?
+        .pubdata_sending_mode
+        == PubdataSendingMode::Custom
+    {
+        anyhow::ensure!(
+            components.contains(&Component::DADispatcher),
+            "Custom pubdata sending mode requires the DA dispatcher component to be enabled"
+        );
```

Comment:
Now that I've looked at this condition again, it doesn't look right; sorry for not noticing previously. This essentially means that if a main node is run in a distributed way (as we do for our main node), then, if all node instances are provided with the same config (which looks reasonable), they all must have the dispatcher component, which doesn't look reasonable; AFAIU, the dispatcher is supposed to be a singleton. It probably makes sense to reverse the condition; i.e., check for `components.contains(&Component::DADispatcher)` at the top level `if` and check `pubdata_sending_mode` inside.

Reply:
I actually realized that it's much more convenient for tests and local development to consider the `da_dispatcher` a default component, and skip its initialization if the pubdata mode is not Custom. Otherwise, it might be a frequent case that users would start a server without a `da_dispatcher` and spend a lot of time figuring out why the server doesn't work in Validium mode. So I did what you suggested, but I write a `warn!` message instead of calling the `ensure!`
Let me know if that doesn't seem like a right approach to you